Fix an issue with finding the Events module for namespaced models

babney opened the pull request:
Hi, I ran into an issue when trying to use this with a model nested under a namespace. Things seemed to work in general, even for the model which raised the issue, but when running our unit test suite, the first test (of any kind) would always fail with `NameError: uninitialized constant Namespace::ModelName`. This appears to fix that. Let me know if you want a test case for this; when I get some time I can probably recreate the problem in a unit test.

babney commented:
> Hi, I ran into an issue when trying to use this with a model nested under a namespace. Things seemed to work in general, even for the model which raised the issue, but when running our unit test suite, the first test (of any kind) would always fail with `NameError: uninitialized constant Namespace::ModelName`. This appears to fix that. Let me know if you want a test case for this; when I get some time I can probably recreate the problem in a unit test.

It occurs to me (now that I stepped away from this) that this might need to be const_get('Events', false) or it'll look up the chain to find some other module named Events. I'll give that a try when I get back to this.

guillermoap commented:
hi @babney thanks for the PR! 
Just to make sure I understand the issue, the problem is only under specs? How does the file structure for the model look?
Are you using the [modularized](https://github.com/rootstrap/active_outbox?tab=readme-ov-file#modularized-outbox-mappings) approach?

A test case for this would be good either way, but first I want to make sure that this is not a config issue (it seems it's not).

babney commented:
> hi @babney thanks for the PR! Just to make sure I understand the issue, the problem is only under specs? How does the file structure for the model look? Are you using the [modularized](https://github.com/rootstrap/active_outbox?tab=readme-ov-file#modularized-outbox-mappings) approach?
> 
> A test case for this would be good either way, but first I want to make sure that this is not a config issue (it seems it's not).

OK, as soon as I did this it failed for me in the same way. I'll find some time to keep looking into this.

babney commented:
> hi @babney thanks for the PR! Just to make sure I understand the issue, the problem is only under specs? How does the file structure for the model look? Are you using the [modularized](https://github.com/rootstrap/active_outbox?tab=readme-ov-file#modularized-outbox-mappings) approach?
> 
> A test case for this would be good either way, but first I want to make sure that this is not a config issue (it seems it's not).

I am not currently modularizing things, it's all under one Outbox model. And it does seem to only happen in tests; I don't see any errors when fiddling around in the console.

The model that's failing is in a namespace like $RAILS_ROOT/app/models/namespace/model.rb

e.g.

```
class Namespace::Model < ApplicationRecord
  include ActiveOutbox::Outboxable
  ...
end
```

In any case, I'm going to spend some more time with this and see if I can reproduce it outside of our app.

babney commented:
> > hi @babney thanks for the PR! Just to make sure I understand the issue, the problem is only under specs? How does the file structure for the model look? Are you using the [modularized](https://github.com/rootstrap/active_outbox?tab=readme-ov-file#modularized-outbox-mappings) approach?
> > A test case for this would be good either way, but first I want to make sure that this is not a config issue (it seems it's not).
> 
> I am not currently modularizing things, it's all under one Outbox model. And it does seem to only happen in tests; I don't see any errors when fiddling around in the console.
> 
> The model that's failing is in a namespace like $RAILS_ROOT/app/models/namespace/model.rb
> 
> e.g.
> 
> ```
> class Namespace::Model < ApplicationRecord
>   include ActiveOutbox::Outboxable
>   ...
> end
> ```
> 
> In any case, I'm going to spend some more time with this and see if I can reproduce it outside of our app.

Also, I can already see this is NOT the solution, it's just picking up another Events module in our app, "above" where it should be looking, so it doesn't raise the error, but it's not finding a Namespace::Events module.

babney commented:
Redid this, and it's a smaller change now. And I get what's happening in our app, and why it failed. Before this change, it would find another module in our app called Events, at the top level, and thus not define Events on the namespace. This change makes it look only inside `module_parent`, and not above it. Everything still seems to work but I think I know how to create an otherwise-failing test case now. I'm at the end of my day here so I'll get on that tomorrow morning.

babney commented:
OK, have a look here. Try leaving out my change in the Outboxable module, but leave that Events module there in the test setup. All I've done is create a module under Object, but it breaks the entire test suite, unless the change in my previous commit is made.

It appears to me that the intent in that `const_defined?` call is to find the module inside of `module_parents`. If that's the case, I think this is all you need to avoid this problem.

Another option that might work, and one you might want to consider anyway, if my change differs from intended behavior: Instead of creating modules called Events there, why not OutboxableEvents? That is much less likely to collide with a module that exists in the application's object space already.

guillermoap commented:
> Instead of creating modules called Events there, why not OutboxableEvents
I think this is the better approach, the main idea for this gem is to be a drop in enhancement that doesn't mess with how you're doing things. So not overriding a possible constant in an app makes sense to me.
If you change that const name to `OutboxableEvents` does it fix your issue?

babney commented:
> ActiveOutboxEvents

Finally got to get back in and give this a try, and it still fails, but again, if I pass false as the second argument to `const_defined?` it works no matter what I call the module.

babney commented:
Some more info for you: I ran my tests with this gem vendored and some puts-debugging.

I changed the name of the Events module to ActiveOutboxEvents. Then, before lib/active_outbox/outboxable.rb:19, I added:

`puts "defining: #{module_parent.name}::ActiveOutboxEvents.const_set(#{const_name}, #{const_name}#{namespace.blank? ? '' : '.'}#{namespace})"`

With the change I made, and with some censorship of our internal model names... I get the following with my change in place:

With the change I made to the `const_defined?` call:

```
defining: Object::ActiveOutboxEvents.const_set(MODEL_ONE_CREATED, MODEL_ONE_CREATED)                                 
defining: Object::ActiveOutboxEvents.const_set(MODEL_ONE_UPDATED, MODEL_ONE_UPDATED)       
defining: Object::ActiveOutboxEvents.const_set(MODEL_ONE_DESTROYED, MODEL_ONE_DESTROYED)     
defining: NamespaceOne::ActiveOutboxEvents.const_set(MODEL_TWO_CREATED, MODEL_TWO_CREATED.NAMESPACE)
defining: NamespaceOne::ActiveOutboxEvents.const_set(MODEL_TWO_UPDATED, MODEL_TWO_UPDATED.NAMESPACE)    
defining: NamespaceOne::ActiveOutboxEvents.const_set(MODEL_TWO_DESTROYED, MODEL_TWO_DESTROYED.NAMESPACE)  
defining: Object::ActiveOutboxEvents.const_set(MODEL_THREE_CREATED, MODEL_THREE_CREATED)                
defining: Object::ActiveOutboxEvents.const_set(MODEL_THREE_UPDATED, MODEL_THREE_UPDATED)       
defining: Object::ActiveOutboxEvents.const_set(MODEL_THREE_DESTROYED, MODEL_THREE_DESTROYED)     
defining: Object::ActiveOutboxEvents.const_set(MODEL_FOUR_CREATED, MODEL_FOUR_CREATED)             
defining: Object::ActiveOutboxEvents.const_set(MODEL_FOUR_UPDATED, MODEL_FOUR_UPDATED)       
defining: Object::ActiveOutboxEvents.const_set(MODEL_FOUR_DESTROYED, MODEL_FOUR_DESTROYED)     
defining: Object::ActiveOutboxEvents.const_set(MODEL_FIVE_CREATED, MODEL_FIVE_CREATED)  
defining: Object::ActiveOutboxEvents.const_set(MODEL_FIVE_UPDATED, MODEL_FIVE_UPDATED)       
defining: Object::ActiveOutboxEvents.const_set(MODEL_FIVE_DESTROYED, MODEL_FIVE_DESTROYED)  
```

After removing that change:

```
defining: Object::ActiveOutboxEvents.const_set(MODEL_ONE_CREATED, MODEL_ONE_CREATED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_ONE_UPDATED, MODEL_ONE_UPDATED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_ONE_DESTROYED, MODEL_ONE_DESTROYED)
E

Error:
NamespaceTwo::ModelSixTest#test_should_return_something:
NameError: uninitialized constant NamespaceOne::ActiveOutboxEvents
    app/models/namespace_one/model_two.rb:2:in `include'
    app/models/namespace_one/model_two.rb:2:in `<class:ModelTwo>'
    app/models/namespace_one/model_two.rb:1:in `<top (required)>'


rails test test/models/namespace_two/model_six_test.rb:4

defining: Object::ActiveOutboxEvents.const_set(MODEL_THREE_CREATED, MODEL_THREE_CREATED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_THREE_UPDATED, MODEL_THREE_UPDATED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_THREE_DESTROYED, MODEL_THREE_DESTROYED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_FOUR_CREATED, MODEL_FOUR_CREATED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_FOUR_UPDATED, MODEL_FOUR_UPDATED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_FOUR_DESTROYED, MODEL_FOUR_DESTROYED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_FIVE_CREATED, MODEL_FIVE_CREATED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_FIVE_UPDATED, MODEL_FIVE_UPDATED)
defining: Object::ActiveOutboxEvents.const_set(MODEL_FIVE_DESTROYED, MODEL_FIVE_DESTROYED)
```

So I can understand a little better now... Is the intent to create a _single_ Events/ActiveOutboxEvents module under Object (given most Rails models are not namespaced), or is the intent to create one under Object for non-namespaced models, and another under each model namespace that is encountered?

guillermoap commented:
> or is the intent to create one under Object for non-namespaced models, and another under each model namespace that is encountered?

This is the intended approach, for more flexibility in supporting namespacing. I'll add it to the README since it's a worthwhile explanation.

I think changing the name to ActiveOutboxEvents is still desired. Adding `false` to `const_get` makes sense, we even had to do it for the [OUTBOX_MODEL](https://github.com/rootstrap/active_outbox/blob/53caa02a437d0a392b1eddaf06e90011c109bf2a/lib/active_outbox/outboxable.rb#L60) constant.

let me know if you need a hand to make this change, because I want you to make the contribution given how much time you've put into this, but I also want to fix this issue.

babney commented:
Hi, finding it hard to find time to work on this mixed into my usual work day... I tried finding a way to create a test that wouldn't create new constants, etc, and really test the problem by itself, but I couldn't find a way to make a failing test... However, I found that if you define these two test classes in reverse order the entire suite breaks, and this is why the test suite in our app broke, and also why I couldn't break it with a simple test. If you define all this on a top-level class like FakeModel _first_, nothing goes wrong, and it'll find that top-level Events/ActiveOutboxEvents module on Object even for namespaced models. But, define the namespaced model first, and it all goes kaboom.

babney commented:
> Hi, finding it hard to find time to work on this mixed into my usual work day... I tried finding a way to create a test that wouldn't create new constants, etc, and really test the problem by itself, but I couldn't find a way to make a failing test... However, I found that if you define these two test classes in reverse order the entire suite breaks, and this is why the test suite in our app broke, and also why I couldn't break it with a simple test. If you define all this on a top-level class like FakeModel _first_, nothing goes wrong, and it'll find that top-level Events/ActiveOutboxEvents module on Object even for namespaced models. But, define the namespaced model first, and it all goes kaboom.

This is why it's going to be difficult to write a single failing test, too... It would have to run before the outbox_models.rb file is required.

guillermoap commented:
> define the namespaced model first, and it all goes kaboom.

Interesting, if you want I can take over this once I have some free time. Thanks for all the input and feedback. If you wish we can go ahead with the ActiveOutboxEvents, and const_get(false) changes and I'll write some specs later. So we can finally fix this issue for you.

babney commented:
> > define the namespaced model first, and it all goes kaboom.
> 
> Interesting, if you want I can take over this once I have some free time. Thanks for all the input and feedback. If you wish we can go ahead with the ActiveOutboxEvents, and const_get(false) changes and I'll write some specs later. So we can finally fix this issue for you.

That would be fantastic. I'll use it as-is vendored but I'll watch for a point release from you folks.

Thanks for your help on this and to everyone on the team there for making this gem in the first place!

guillermoap commented:
no worries! Thank you for taking the time to bring this up and trying to make a PR. I'll mark this as closed then, but any suggestions or other comments you have feel free to open an issue.

guillermoap commented:
Will address this changes in another PR
